docs/design-docs/08-package-variant.md

```diff
@@ -0,0 +1,450 @@
+not needed. Rather, when the floating tag is moved to a new tagged revision of
+the upstream package, the package revision controller will notice and
+automatically propose and update to that revision. For example, the upstream
+package author may designate three floating tags: stable, beta, and alpha. The
+upstream package author can move these tags to specific revisions, and any
+PackageVariant resource tracking them will propose updates to their downstream
+packages.
+
+
+
+### Adoption and Deletion Policies
+When a `PackageVariant` resource is created, it will have a particular
+repository and package name as the downstream. The adoption policy controls
+whether the package variant controller takes over an existing package with that
+name, in that repository.
+
+Analogously, when a `PackageVariant` resource is deleted, a decision must be
+made about whether or not to delete the downstream package. This is controlled
+by the deletion policy.
+
+### Deletion Policy
+
+## Fan Out of Variant Generation
+
+TODO(johnbelamaric): Describe `PackageVariantSet`.
+
+## Example Use Cases
+
+- describe scenarios and the `PackageVariant`, `PackageVariantSet` resources that
+  would solve the scenarios
+
+### Automatically Organizational Customization of an External Package
+We can use a `PackageVariant` to provide basic changes that are needed when
+importing a package from an external repository. For example, suppose we have
+our own internal registry from which we pull all our images. When we import an
+upstream package, we want to modify any images listed in the upstream package to
+point to our registry. Additionally, we have a policy that all Pods must have a
+`chargeback-code` label, so we want to add that as a validating policy to the
+package.
+
+In a manual CLI-based workflow, we would accomplish this as follows.
+...
+
+With `PackageVariant`, we instead create the following resources:
+...
+
+### Creating a Namespace Per Tenant
+
+### Customizing A Workload By Region
+
+### Customizing A Workload By Cluster and Region
+
+### Customizing A Workload By Environment
+
+## Detailed Design
+
+### `PackageVariant` API
+
+#### Basic API
+
+- Upstream
+- Downstream
+- Annotations
+- Labels
+- Adoption and Deletion Policy
+
+#### Package Context Injection
+
+#### KRM Function Pipeline
+
+#### Configuration Injection
+
+Configuration injection is controlled by a combination of in-package resources
+with annotations, and *injection selectors* defined on the PackageVariant
+resource.
+
+Injection selectors are defined in the `spec.injectionSelectors` field of the
+`PackageVariant`. This field is an ordered array of structs containing group,
+version, kind, and name. Only the name is required.
+
+The injection process works as follows:
+
+1. The controller will examine all in-package resources, looking for those with
+   an annotation named `kpt.dev/config-injection`, with one of the following
```

Comment:
I prefer to keep it as short as possible, but whatever you/Morten/Sunil/Justin think works.

Reply:
For the finalizers, we have just a convention now to reuse the group of the resource type managed by the controller. I think we can use the short form for well-known annotations/labels. Hopefully the number of those will be relatively small.

Reply:
At least I think this should apply to annotations/labels that are used for general mechanisms in Porch. If it is something that is specific to a controller (like PackageVariantSet keeping track of PackageVariants), we probably want to use something more specific to the controller.

Reply:
So `kpt.dev` or `porch.kpt.dev`?

Reply:
I think I prefer just `kpt.dev`. Including the name of the product (in this case Porch) in the group doesn't seem necessary and it will just be confusing if we want to use the annotation outside of Porch or if we at some point change the name. It is also shorter.

Reply:
ok, will do